### ISSUE TITLE
Addressing Known High CVE Vulnerabilities in Spring

### ISSUE THREAD

**Issue by davidgoate** [spring-projects/spring-boot#9997]
# Background

Let me start by saying I love spring/spring boot. I do most of my day to day work using it and intend to continue using it often. I appreciate the hard work that goes on and the productivity gains that I feel I get from using it.

## Fine, but down to business
I use the OWASP maven plugin to perform a project dependency scan during the build. I had intended to set this up to only fail the build on "high" priority issues using scores mentioned [in the NVD](https://nvd.nist.gov/vuln-metrics/cvss). 

Basically for scores of 7 or above I had intended to fail my build. 

## Plugin Configuration
 ```
<properties>
 <owasp.dependency.check.cve.cveValidForHours>48</owasp.dependency.check.cve.cveValidForHours>
<!-- Do not fail the build, numbers are 0-10 so this won't even fail the build for now-->
<owasp.dependency.check.cvss.limit>11</owasp.dependency.check.cvss.limit>
</properties>

<plugin>
        <groupId>org.owasp</groupId>
        <artifactId>dependency-check-maven</artifactId>
        <version>1.4.5</version>
        <executions>
          <execution>
            <goals>
              <goal>check</goal>
            </goals>
            <configuration>
              <skip>${nvd.skip}</skip>
              <cveValidForHours>${owasp.dependency.check.cve.cveValidForHours}</cveValidForHours>
              <assemblyAnalyzerEnabled>false</assemblyAnalyzerEnabled>
            </configuration>
          </execution>
        </executions>
      </plugin>
```

# Identified Vulnerabilities

I was then actually a little surprised (naively?) that my build failed using the latest/current GA stable version of boot (1.5.6.RELEASE). 

These are the vulnerabilities identified (not all are high severity).

* jackson-annotations-2.8.0.jar (com.fasterxml.jackson.core:jackson-annotations:2.8.0, cpe:/a:fasterxml:jackson:2.8.0) : CVE-2016-7051
* jackson-core-2.8.9.jar (com.fasterxml.jackson.core:jackson-core:2.8.9, cpe:/a:fasterxml:jackson:2.8.9) : CVE-2016-7051
* tomcat-embed-core-8.5.16.jar (cpe:/a:apache:tomcat:8.5.16, org.apache.tomcat.embed:tomcat-embed-core:8.5.16) : CVE-2017-6056, CVE-2016-6325, CVE-2016-5425
* spring-boot-starter-data-redis-1.5.6.RELEASE.jar (cpe:/a:pivotal_software:redis:1.5.6, org.springframework.boot:spring-boot-starter-data-redis:1.5.6.RELEASE) : CVE-2015-8080, CVE-2015-4335, CVE-2013-7458

# Thoughts

I am fully aware that these vulnerabilities aren't in spring boot itself and that they can't (necessarily) be fixed by spring developers, so I'm not raising this issue in some attempt to hope that Spring developers will be able to magically make them go away. Perhaps some are even false positives on the libs...

I guess what I am hoping for is consideration/discussion about the following:

1. Whether some libraries could be updated to later versions that don't contain these flaws. I know that this could be complex due to different dependencies requiring different versions of a 3rd party lib within boot.
1. If these can't be updated/fixed, perhaps a section could be added to the documentation of known vulnerabilities and how to practically address them or assess whether it's relevant to you. E.g. https://nvd.nist.gov/vuln/detail/CVE-2016-7051 suggests the Jackson one would only be an issue if you're using the XMLMapper. I presume a lot of people using Jackson only utilise the JSON mappings and not XML ones, so may be safe even using a vulnerable version of the lib. 
1. Perhaps it turns out spring boot isn't vulnerable to some of these flaws due to those particular features not being used or exposed by default. Again, perhaps some research/documentation could be done here to help users understand.
1. Perhaps part of the spring boot build process could integrate this OWASP check and include reports in the documentation automatically to give people more visibility of known issues in the latest version(s).

I guess, overall, I'm feeling that "ignorance isn't an excuse". So, before I had run this tool I had no idea about these known vulnerabilities in the application(s) that I am shipping to production. Now that I know about them I'm probably in a better position, as I can assess the risks, look at mitigation techniques and generally feel slightly more comfortable about the apps I'm shipping. *But* the point is, before I even write a line of code, just using a fresh project from start.spring.io I am starting from a baseline where my app has approx 7 known vulnerabilities - some ranked as high severity.

Of course there can be plenty more unknown vulnerabilities in these dependencies and I can introduce a bunch more through other libs I add. An indeed perhaps the biggest attack vector of all probably may be the actual business logic/code that I write rather than these well tested battle hardened libs. However, I feel i sense of responsibility to my employers, myself, my users and the community to raise this issue for discussion. Knowingly shipping apps with issues that rank as high on the NVD CVE [example](https://nvd.nist.gov/vuln/detail/CVE-2016-6325) might make some users uncomfortable.

I know spring is used heavily in big enterprises and I haven't (yet) performed similar tests on JEE libs or other frameworks so I'm not able to comment on how they compare and contrast on known vulnerabilities. Although it's stating the obvious, I feel the need to say it, I am aware no software is bug free. Security issues will always exist and as new research comes along new issues will be found in code previously considered secure (new & old alike). I don't intended to make this a finger pointing exercise of "hey, look spring is insecure". Indeed, as mentioned above, these issues are in other dependencies that spring happens to bring in. But just as I should be accountable for my decisions to bring in external dependencies onto a project I think it'd be nice if this aspect could be improved in Spring Boot.

I'd really like feedback/suggestions/discussion about this.

**Comment by wilkinsona** [comment]
Unfortunately, in my experience and the experience of Pivotal's security team, automated security scans are of little value as they generate an overwhelming number of false positives.

In this case, all 7 of the vulnerabilities are false positives:

- CVE-2017-6056 is specific to Debian's Tomcat package
- CVE-2016-7051 has been fixed in Jackson 2.8.4. `jackson-annotations` only contains annotations and is unaffected
- CVE-2016-6325 is specific to RHEL's Tomcat package
- CVE-2016-5425 is specific to RHEL's Tomcat package
- CVE-2015-8080 is a vulnerability in Redis, not Spring Data Redis
- CVE-2015-4335 is a vulnerability in Redis, not Spring Data Redis
- CVE-2013-7458 is a vulnerability in Redis, not Spring Data Redis

Unfortunately, I don't think there's anything that we can realistically do in Spring Boot to address this. Tools like the OWASP dependency checker generate so many false positives that we simply don't have time to document them and explain why they are false positives.

Please rest assured that we do take security seriously. We regularly update our dependencies on other Spring projects and third-party libraries to keep up-to-date. Vulnerabilities in Spring projects are listed on the [Pivotal Security site](https://pivotal.io/security) which I would encourage you to keep an eye on. On the rare occasion where there has been a vulnerability found in a Spring project, we have coordinated with that project to update Spring Boot as soon as a release that addresses the vulnerability is available. Typically the release announcement (on spring.io/blog) will mention when a release addresses a vulnerability.

**Comment by davidgoate** [comment]
@wilkinsona thanks for the detailed and fast response.
